MuonAnalysis/MuonAssociators/plugins/MuonFSRProducer.cc

```diff
@@ -0,0 +1,217 @@
+                                  const pat::PackedCandidateCollection& pfcands,
+                                  const double& isoConeMax,
+                                  const double& isoConeMin) const;
+
+  // ----------member data ---------------------------
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> pfcands_;
+  const edm::EDGetTokenT<pat::ElectronCollection> electrons_;
+  const edm::EDGetTokenT<edm::View<reco::Muon>> muons_;
+  float ptCut;
+  float etaCut;
+  float photonPtCut;
+  float drEtCut;
+  float isoCut;
+};
+
+void MuonFSRProducer::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  using namespace std;
+
+  edm::Handle<pat::PackedCandidateCollection> pfcands;
+  iEvent.getByToken(pfcands_, pfcands);
+  edm::Handle<edm::View<reco::Muon>> muons;
+  iEvent.getByToken(muons_, muons);
+  edm::Handle<pat::ElectronCollection> electrons;
+  iEvent.getByToken(electrons_, electrons);
+
+  auto fsrPhotons = std::make_unique<std::vector<pat::GenericParticle>>();
+  // loop over all muons
+  for (auto muon = muons->begin(); muon != muons->end(); ++muon) {
+    int photonPosition = -1;
+    double distance_metric_min = -1;
+    // minimum muon pT
+    if (muon->pt() < ptCut)
+      continue;
+    // maximum muon eta
+    if (abs(muon->eta()) > etaCut)
+      continue;
+
+    // for each muon, loop over all pf cadidates
+    for (auto iter_pf = pfcands->begin(); iter_pf != pfcands->end(); iter_pf++) {
+      auto const& pc = *iter_pf;
+
+      // consider only photons
+      if (abs(pc.pdgId()) != 22)
+        continue;
+      // minimum pT cut
+      if (pc.pt() < photonPtCut)
+        continue;
+
+      // eta requirements
+      if (abs(pc.eta()) > 1.4 and (abs(pc.eta()) < 1.6))
+        continue;
+      if (abs(pc.eta()) > 2.5)
+        continue;
+
+      // 0.0001 < DeltaR(photon,muon) < 0.5 requirement
+      double dRPhoMu = deltaR(muon->eta(), muon->phi(), pc.eta(), pc.phi());
+      if (dRPhoMu < 0.0001)
+        continue;
+      if (dRPhoMu > 0.5)
+        continue;
+
+      bool skipPhoton = false;
+      bool closest = true;
+
+      for (auto othermuon = muons->begin(); othermuon != muons->end(); ++othermuon) {
+        if (othermuon->pt() < ptCut or abs(othermuon->eta()) > etaCut or muon == othermuon)
+          continue;
+        double dRPhoMuOther = deltaR(othermuon->eta(), othermuon->phi(), pc.eta(), pc.phi());
+        if (dRPhoMuOther < dRPhoMu) {
+          closest = false;
+          break;
+        }
+      }
+      if (!closest)
+        continue;
+
+      // Check that is not in footprint of an electron
+      pat::PackedCandidateRef pfcandRef = pat::PackedCandidateRef(pfcands, iter_pf - pfcands->begin());
+
+      for (auto electrons_iter = electrons->begin(); electrons_iter != electrons->end(); ++electrons_iter) {
+        for (auto itr = electrons_iter->associatedPackedPFCandidates().begin();
```

Comment:
This does not work. `associatedPackedPFCandidates()` returns a temporary object

https://github.com/cms-sw/cmssw/blob/a41ba6cb802a727726f33a31be291ca441534016/DataFormats/PatCandidates/interface/Electron.h#L260

so the calls to `electrons_iter->associatedPackedPFCandidates().begin()` and `electrons_iter->associatedPackedPFCandidates().begin()` are actually returning iterators to two different containers (both of which are gone during the iteration of the loop).

This was found by the address sanitizer
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc820/CMSSW_11_0_ASAN_X_2019-10-18-2300/pyRelValMatrixLogs/run/250202.118_ProdTTbar_13_pmx25ns+ProdTTbar_13UP18+DIGIPRMXUP18_PROD_PU25+RECOPRMXUP18PROD_PU25+NANOEDMMC2018_PROD/step4_ProdTTbar_13_pmx25ns+ProdTTbar_13UP18+DIGIPRMXUP18_PROD_PU25+RECOPRMXUP18PROD_PU25+NANOEDMMC2018_PROD.log#/

Reply:
shouldn't it loop forever then?

Reply:
It's actually undefined behavior. I'm guessing that the two temporaries just happen to be made at the same point of the stack. This causes a luck accident that it just happens to work for now.

Reply:
ok then we may experience random crashes in large production.
@peruzzim I'm not sure if the nanov6 campaign was already submitted, but I'd suggest to fix this (just having a local copy of the vector)

Reply:
I think a possibly better solution would be to switch to using ranged for loop
```C++
for(auto const& cand: electrons_iter->associatedPackedPFCandidates()) {
```
The language guarantees that the temporary container returned from the call to `electrons_iter->associatedPackedPFCandidates()` will be kept around for the lifetime of the for loop.

Reply:
implemented this solution in #28221